### PR TITLE
Add the threshold control whether the HTTP response should be compressed.

### LINKED ISSUE #7660: io.netty.handler.codec.http.HttpContentCompressor need compress threshold
### Expected behavior
the response body is compressed, when the response body exceeds the threshold
### Actual behavior
no threshold can be set
### Steps to reproduce
-
### Minimal yet complete reproducer code (or URL to code)
-
### Netty version
4.1.16.Final
### JVM version (e.g. `java -version`)
java 1.8
### OS version (e.g. `uname -a`)
-

### PR DESCRIPTION
**Pull request by teaey**
…mpressed.

Motivation:

Add the threshold control whether the HTTP response should be compressed.

Modifications:

Add filed io.netty.handler.codec.http.HttpContentCompressor#compressThreshold that control whether the HTTP response should be compressed.

Result:

Fixes #7660.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 1]
```diff
@@ -33,6 +33,7 @@
     private final int windowBits;
     private final int memLevel;
     private ChannelHandlerContext ctx;
+    private final int compressThreshold;
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: nit: can you move this to the other final variables.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 2]
```diff
@@ -52,7 +53,7 @@ public HttpContentCompressor() {
      *        compression level is {@code 6}.
      */
     public HttpContentCompressor(int compressionLevel) {
-        this(compressionLevel, 15, 8);
+        this(compressionLevel, 15, 8,0);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: add whitespace after `,`

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 3]
```diff
@@ -73,8 +74,11 @@ public HttpContentCompressor(int compressionLevel) {
      *        state.  {@code 1} uses minimum memory and {@code 9} uses maximum
      *        memory.  Larger values result in better and faster compression
      *        at the expense of memory usage.  The default value is {@code 8}
+     * @param compressThreshold
+     *        The response body is compressed,
+     *        when the response body exceeds the threshold
      */
-    public HttpContentCompressor(int compressionLevel, int windowBits, int memLevel) {
+    public HttpContentCompressor(int compressionLevel, int windowBits, int memLevel, int compressThreshold) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: We need to keep the existing constructor to not break the API. Please just create an extra constructor and delegate to it.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 4]
```diff
@@ -73,8 +74,11 @@ public HttpContentCompressor(int compressionLevel) {
      *        state.  {@code 1} uses minimum memory and {@code 9} uses maximum
      *        memory.  Larger values result in better and faster compression
      *        at the expense of memory usage.  The default value is {@code 8}
+     * @param compressThreshold
+     *        The response body is compressed,
+     *        when the response body exceeds the threshold
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: Explain that `<= 0` will disable this.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 5]
```diff
@@ -107,6 +112,12 @@ protected Result beginEncode(HttpResponse headers, String acceptEncoding) throws
             return null;
         }
 
+        if(this.compressThreshold > 0) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: add space after `if`

**Comment by teaey** [comment]
@normanmaurer  I suppose using strategy pattern which is better.
I will introduced a method eg:

```java
boolean needCompress(Response response);
```

**Comment by normanmaurer** [comment]
@teaey please add a unit test as well.

**Comment by teaey** [comment]
@normanmaurer Two methods are added to the HttpContentCompressorTest.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 6]
```diff
@@ -131,6 +164,18 @@ protected Result beginEncode(HttpResponse headers, String acceptEncoding) throws
                         wrapper, compressionLevel, windowBits, memLevel)));
     }
 
+    /**
+     * For customization should implement this method that
+     * performs the necessary business logic.
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: @teaey please reword to:

Returns {@code true} if the compression should be used. May be override to change default behaviour

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 7]
```diff
@@ -131,6 +164,18 @@ protected Result beginEncode(HttpResponse headers, String acceptEncoding) throws
...
+     */
+    protected boolean needCompress(HttpResponse response) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: rename to `useCompression(...)`

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 8]
```diff
@@ -75,6 +76,33 @@ public HttpContentCompressor(int compressionLevel) {
...
+     *        Larger values result in better compression at the expense of
+     *        memory usage.  The default value is {@code 15}.
+     * @param memLevel
+     *        How much memory should be allocated for the internal compression
+     *        state.  {@code 1} uses minimum memory and {@code 9} uses maximum
+     *        memory.  Larger values result in better and faster compression
+     *        at the expense of memory usage.  The default value is {@code 8}
+     * @param compressThreshold
+     *        The response body is compressed,
+     *        If the response body exceeds the threshold.
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: nit: s/If/if/

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 9]
```diff
@@ -75,6 +76,33 @@ public HttpContentCompressor(int compressionLevel) {
...
+     *        Default value is {@code 0} that means all responses will be compressed.
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: change to: `{@code 0}` will enable compression for all responses.

**Comment by teaey** [comment]
@normanmaurer All changed.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 10]
```diff
@@ -107,6 +136,10 @@ protected Result beginEncode(HttpResponse headers, String acceptEncoding) throws
             return null;
         }
 
+        if (!useCompression(headers)) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: @teaey sorry one last nit... I think this should be the first thing to check as we not need to even compare headers if it is false. Please move it before the header check

**Comment by teaey** [comment]
@normanmaurer I think so.

**Comment by normanmaurer** [comment]
@ejona86 @Scottmitch @carl-mastrangelo @trustin WDYT ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 11]
```diff
@@ -131,6 +164,18 @@ protected Result beginEncode(HttpResponse headers, String acceptEncoding) throws
                         wrapper, compressionLevel, windowBits, memLevel)));
     }
 
+    /**
+     * Returns {@code true} if the compression should be used.
+     * May be override to change default behaviour.
+     */
+    protected boolean useCompression(HttpResponse response) {
+        if (this.compressThreshold > 0) {
+            return response instanceof HttpContent &&
```
Comment (ejona86) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: This looks like it only works properly for a FullHttpResponse. If a HttpResponse was sent separately than a HttpContent(s), then this won't compress. Is that intended?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 12]
```diff
@@ -75,6 +76,33 @@ public HttpContentCompressor(int compressionLevel) {
...
+     *        The base two logarithm of the size of the history buffer.  The
+     *        value should be in the range {@code 9} to {@code 15} inclusive.
+     *        Larger values result in better compression at the expense of
+     *        memory usage.  The default value is {@code 15}.
+     * @param memLevel
+     *        How much memory should be allocated for the internal compression
+     *        state.  {@code 1} uses minimum memory and {@code 9} uses maximum
+     *        memory.  Larger values result in better and faster compression
+     *        at the expense of memory usage.  The default value is {@code 8}
+     * @param compressThreshold
```
Comment (ejona86) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: sizeThreshold? payloadSizeThreshold? compressThreshold just makes me question what it means (e.g., threshold of compression ratio?).

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 13]
```diff
@@ -32,6 +32,7 @@
     private final int compressionLevel;
     private final int windowBits;
     private final int memLevel;
+    protected final int compressThreshold;
```
Comment (ejona86) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: Why is this protected?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 14]
```diff
@@ -91,6 +119,7 @@ public HttpContentCompressor(int compressionLevel, int windowBits, int memLevel)
         this.compressionLevel = compressionLevel;
         this.windowBits = windowBits;
         this.memLevel = memLevel;
+        this.compressThreshold = (compressThreshold < 0) ? 0 : compressThreshold;
```
Comment (ejona86) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: I'd feel better with throwing if negative.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java (hunk 13 above)
Comment (teaey) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: Hi @ejona86. 

Users can implement the subclass to overwrite the useCompression method, which may access the compressThreshold.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java (hunk 14 above)
Comment (teaey) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: Yes, this is the same as other fields.
@normanmaurer What's your opinion?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java (hunk 12 above)
Comment (teaey) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: 👍 
What about contentSizeThreshold?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java (hunk 11 above)
Comment (teaey) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: This is in line with the original class.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java (hunk 12 above)
Comment (ejona86) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: SGTM

**Comment by teaey** [comment]
@ejona86 Please review again.

**Comment by ejona86** [comment]
@teaey, oh, I had been waiting for a reply on one of my comments. Since the most recent commit wasn't done in a separate commit I couldn't easily tell what changes you made.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (06dcca1dbcc048b6ce7d954e4ecdf1d3ed6af034).

@teaey thanks!

**Pull request closed without merging**